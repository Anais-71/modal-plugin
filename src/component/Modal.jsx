import React, { useEffect, useRef } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faXmark } from '@fortawesome/free-solid-svg-icons'
import PropTypes from 'prop-types'
import './modal.css'

/**
 * Modal component for displaying a popup dialog with accessibility enhancements.
 *
 * @param {Object} props - The properties for the Modal component.
 * @param {Function} props.onClose - Function to call when the modal is closed.
 * @param {string} [props.title] - Optional title of the modal dialog.
 * @param {string} [props.message] - Optional message to display in the modal dialog.
 * @param {Array<{ label: string, onClick: Function }>} props.actions - Array of action buttons, each with a label and an onClick handler.
 *
 * @returns {JSX.Element} The rendered modal component.
 */
function Modal({ onClose, title, message, actions }) {
  // Reference to the modal container for managing focus
  const modalRef = useRef(null)

  /**
   * Sets focus on the modal container element when the component mounts.
   */
  useEffect(() => {
    modalRef.current.focus()
  }, [])

  /**
   * Handles keyboard interactions for accessibility.
   * Closes the modal when the "Escape" key is pressed.
   *
   * @param {KeyboardEvent} e - The keyboard event.
   */
  const handleKeyDown = (e) => {
    if (e.key === 'Escape') {
      onClose()
    }
  }

  return (
    <div
      className="modal"
      role="dialog"
      aria-modal="true"
      onKeyDown={handleKeyDown}
      ref={modalRef}
      tabIndex="-1"
    >
      <div
        className="modal__popup"
        aria-labelledby="modal-title"
        aria-describedby="modal-message"
      >
        {/* Close button with accessible label */}
        <button
          className="modal__popup--close"
          onClick={onClose}
          aria-label="Close modal"
        >
          <FontAwesomeIcon
            icon={faXmark}
            className="modal__popup--close--icon"
          />
        </button>

        {/* Conditional rendering of title */}
        {title && (
          <h3 id="modal-title" className="modal__popup--title">
            {title}
          </h3>
        )}

        {/* Conditional rendering of message */}
        {message && (
          <p id="modal-message" className="modal__popup--message">
            {message}
          </p>
        )}

        {/* Rendering action buttons provided in the actions prop */}
        <div className="modal__popup--buttons">
          {actions.map((action, index) => (
            <button
              key={index}
              className="modal__popup--buttons--btn"
              onClick={action.onClick}
            >
              {action.label}
            </button>
          ))}
        </div>
      </div>
    </div>
  )
}

Modal.propTypes = {
  onClose: PropTypes.func.isRequired,
  title: PropTypes.string,
  message: PropTypes.string,
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      onClick: PropTypes.func.isRequired,
    }),
  ).isRequired,
}

export default Modal
