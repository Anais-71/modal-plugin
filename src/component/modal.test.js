import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import Modal from './Modal'

describe('Given I use the Modal component', () => {
  const mockOnClose = jest.fn()
  const mockActions = [
    { label: 'Confirm', onClick: jest.fn() },
    { label: 'Cancel', onClick: jest.fn() },
  ]

  beforeEach(() => {
    mockOnClose.mockClear()
    render(
      <Modal
        onClose={mockOnClose}
        title="Test Modal"
        message="This is a test modal message."
        actions={mockActions}
      />,
    )
  })

  /**
   * Test to check if the modal is rendered with a title and message.
   */
  test('Then the modal should render with title and message', () => {
    const title = screen.getByRole('heading', { name: /Test Modal/i })
    const message = screen.getByText(/This is a test modal message./i)
    expect(title).toBeInTheDocument()
    expect(message).toBeInTheDocument()
  })

  /**
   * Test to verify that the close button calls the onClose handler when clicked.
   */
  test('When I click the close button, Then the onClose handler is called', () => {
    const closeButton = screen.getByRole('button', { name: /close modal/i })
    fireEvent.click(closeButton)
    expect(mockOnClose).toHaveBeenCalledTimes(1)
  })

  /**
   * Test to check that action buttons render correctly and trigger the respective handler.
   */
  test('Then the action buttons should render and trigger their onClick handlers', () => {
    const confirmButton = screen.getByText(/Confirm/i)
    const cancelButton = screen.getByText(/Cancel/i)

    expect(confirmButton).toBeInTheDocument()
    expect(cancelButton).toBeInTheDocument()

    fireEvent.click(confirmButton)
    expect(mockActions[0].onClick).toHaveBeenCalledTimes(1)

    fireEvent.click(cancelButton)
    expect(mockActions[1].onClick).toHaveBeenCalledTimes(1)
  })

  /**
   * Test to verify that pressing the Escape key closes the modal.
   */
  test('When I press the Escape key, Then the onClose handler is called once', () => {
    const modal = screen.getByRole('dialog')
    fireEvent.keyDown(modal, { key: 'Escape', code: 'Escape' })
    expect(mockOnClose).toHaveBeenCalledTimes(1)
  })

  /**
   * Test to verify accessibility attributes are correctly set.
   */
  test('Then the modal has correct accessibility attributes', () => {
    const modal = screen.getByRole('dialog')
    expect(modal).toHaveAttribute('aria-modal', 'true')
  })
})

/**
 * Snapshot test for the Modal component.
 */
test('Modal component matches snapshot', () => {
  const { asFragment } = render(
    <Modal
      onClose={() => {}}
      title="Snapshot Test Modal"
      message="Testing snapshot for the modal."
      actions={[{ label: 'OK', onClick: () => {} }]}
    />,
  )
  expect(asFragment()).toMatchSnapshot()
})
