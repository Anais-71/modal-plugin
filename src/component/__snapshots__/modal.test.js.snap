// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-modal="true"
    class="modal"
    role="dialog"
    tabindex="-1"
  >
    <div
      aria-describedby="modal-message"
      aria-labelledby="modal-title"
      class="modal__popup"
    >
      <button
        aria-label="Close modal"
        class="modal__popup--close"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-xmark modal__popup--close--icon"
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            fill="currentColor"
          />
        </svg>
      </button>
      <h3
        class="modal__popup--title"
        id="modal-title"
      >
        Snapshot Test Modal
      </h3>
      <p
        class="modal__popup--message"
        id="modal-message"
      >
        Testing snapshot for the modal.
      </p>
      <div
        class="modal__popup--buttons"
      >
        <button
          class="modal__popup--buttons--btn"
        >
          OK
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
